### DISCUSSION TITLE
Datetime index issue?

### DISCUSSION THREAD

**Discussion by harshhacks** [kernc/backtesting.py#303] [Q&A] [answered]
I've recently been backtesting with a few CSV files and getting this warning everytime I run the backtest:

<ipython-input-75-7760bd4ae4a4>:19: UserWarning: Data index is not datetime. Assuming simple periods, but `pd.DateTimeIndex` is advised.
  bt = Backtest(df_EURUSD, SMACrossover, cash=10000, commission=.000,exclusive_orders=True)

I have a time column which I have converted to DatetimeIndex and set as index but for some reason the backtesting engine adds an index column and uses that for backtesting which results in NaN values for any time-series based results like annual return.

Any ideas on where I might be going wrong?

**Comment by kernc** [comment, marked as answer]
The code uses:
https://github.com/kernc/backtesting.py/blob/e3cccdfc0fdf8f6f1fe0b5c7b84fd5a2b15fab2d/backtesting/backtesting.py#L1047-L1057
https://github.com/kernc/backtesting.py/blob/e3cccdfc0fdf8f6f1fe0b5c7b84fd5a2b15fab2d/backtesting/backtesting.py#L1079-L1082
Apparently, your index fails for:
```py
assert isinstance(df_EURUSD.index, pd.DatetimeIndex), df_EURUSD.index
```

Can you show an example of your data frame?

**Reply by harshhacks** [reply]
![Dataframe](https://user-images.githubusercontent.com/71103927/114093036-9f44ea00-9888-11eb-9f21-21b320934241.PNG)
Thanks for the reply! It looks something like this. I initially assumed my "time" column was not a DatetimeIndex, so I explicitly converted it using the pd.to_datetime() function just to be sure, but kept getting the same warning.

**Reply by kernc** [reply]
Can you show:
```py
>>> df.index.__class__.__mro__
```
The index appears to not be an instance of `pd.DatetimeIndex`. Also try:
```py
df.index = pd.to_datetime(df.index)
```

**Reply by harshhacks** [reply]
Something interesting happened with this. Here is what happened when I ran `df.index.__class__.__mro__`

![Original](https://user-images.githubusercontent.com/71103927/114101525-db317c80-9893-11eb-9c8d-da978e23c1ed.PNG)

When I converted the index to datetime using `df.index = pd.to_datetime(df.index)` , it again added a new index, and then converted that to DatetimeIndex:

![NewDateTimeIndex](https://user-images.githubusercontent.com/71103927/114101557-e2f12100-9893-11eb-86e0-e7adb0cfee28.PNG)

For comparison, these are supposed to be the same dates as:

![Dataframe](https://user-images.githubusercontent.com/71103927/114101790-3c595000-9894-11eb-92c9-d00c26e5bcfe.PNG)

**Reply by kernc** [reply]
Interesting. So your index containing dates was for some reason still a `RangeIndex`, and casting `df.index = pd.to_datetime(df.index)` must have fixed it. Pandas is not bare of odd boogs, that's for sure.

**Reply by harshhacks** [reply]
That's the weird part. It didn't fix it. It just added a new Integer index, and converted that integer index to Datetime. So now, for instance, the first value in the index, supposed to be 2020-01-02 22:01:00 is instead '1970-01-01 00:00:00.00003050

To understand where the 3050 is coming from, the current data frame is sliced from an existing data frame and its starting location was the 3051st row in the original.

**Reply by kernc** [reply]
Oh, I see. Then `pd.to_datetime(df.index)` must have experienced issues interpreting those values as datetime. We can continue to shoot in the dark, or if you can provide a small sample of your data (maybe just 100 rows instead of the full 200k rows) and a reproduction scenario, we can dig deeper.

**Reply by harshhacks** [reply]
GitHub won't let me attach a CSV here. I can maybe upload it to a link and upload? Or maybe zip it up and attach here, whatever you prefer.

**Reply by kernc** [reply]
Rename it to _.txt_.

**Reply by harshhacks** [reply]
[aaares.txt](https://github.com/kernc/backtesting.py/files/6286347/aaares.txt)

Attached the first 100 rows for your reference. Thanks a lot for taking a look!

**Reply by harshhacks** [reply]
For the strategy, I'm just doing something trivial like:

`bt = Backtest(df, SMACrossover, cash=10000, commission=.002,exclusive_orders=True)`

**Reply by kernc** [reply]
This works:
```py
>>> df = pd.read_csv('aaares.txt', index_col='time', parse_dates=True)
>>> df.index.__class__
pandas.core.indexes.datetimes.DatetimeIndex
```

**Reply by harshhacks** [reply]
Thanks a lot! It's working great now.
